Step by step Nodejs REPL tutorial (#2059)
section-engineering-education/engineering-education: articles/step-by-step-on-how-to-user-nodejs-repl-tutorial/index.md

### Introduction

> Node.js `Read -Evaluate-Print-Loop` is an easy-to-use command-line tool for processing Node.js expressions.

Reviewer: Explain briefly about `Read-Evaluate-Print loop`. What's is being `Read`, what is being `Evaluated` and what is `Printed`?

Author: REPL stands for READ-EVALUATE PRINT READ, explained what it does in the subsequent paragraph

> This shell captures the user's JavaScript code inputs, interprets and evaluates the result of this code, displays the result to the screen, and repeats the process till the user quits the shell.  

Reviewer: How is it different from a normal command prompt or a linux terminal?

> It's therefore, important to note that this tool does not require file creation to write codes. It comes with ready to use Node.js development environment.

Reviewer: It would be great if you could give a more in detailed introduction about REPL, why one must choose REPL, what other alternatives do they have, how does it work.

Also, give an overview of what the reader is going to learn in this article.

Refer other section articles to get a better understanding on how to write `Introduction`

Reviewer: Add any link for downloading this command line tool.

### Prerequisites
This tutorial requires:
1. Node.js Installed in your development local machine.
2. Basic knowledge of JavaScript.

> 

Reviewer: Add a `Table of contents` section

### Getting started with REPL

> Enter the following in the terminal:

Reviewer: Before getting started with the tutorial, state the objectives of what is going to be done. This helps readers understand what they're going to learn.

> ```bash
> $ node
> 
> ```

Reviewer: ```suggestion
```bash
$ node
```

```

> Output:

Reviewer: ```suggestion
**Output:**
```

```bash
> 
```

> Entering the command `node` in the terminal starts the REPL command-line tool indicated by the `>` symbol.  

Reviewer: ```suggestion

Entering the command `node` in the terminal starts the REPL command-line tool indicated by the `>` symbol.  
```

This symbol acts as an indicator that JavaScript is ready to read and evaluate your code.  

> We can test the functionality by printing 'hello world' using REPL as shown below:

Reviewer: ```suggestion
We can test the functionality by printing `hello world` using REPL as shown below:
```

```bash
$ node
> let hello = 'hello world';
undefined
> hello
'hello world'  // output
> 

```

> In the script, we declare a variable `hello` and assign the text 'hello world'.  
> Next, we call this variable and press enter, to print the 'hello world' string on the screen. That's it.  

Reviewer: ```suggestion
In the script, we assign the text `hello world` to a variable called `hello`.

Now, on calling this variable `hello`, we get `hello world` as the output on the command shell.
```

Now how do we exit REPL?

> To exit this tool, press `ctrl + Z on your keyboard.  

Reviewer: ```suggestion
To exit this tool, press `Ctrl + Z` on your keyboard.  
```

Output:
```bash
> let hello = 'hello world';
undefined
> hello
'hello world'
> 
(To exit, press Ctrl+C again or Ctrl+D or type .exit)
> 

```
As shown in the output, you can as well use `ctrl+D` or type `.exit`.  

> ### Executing code using REPL.

Reviewer: ```suggestion
### Executing code using REPL
```

As discussed previously, this tool simplifies work by providing a quick way to test Node.js code without creating files.  
> It's also important to note that any valid JavaScript code can be executed using REPL.

>  1. Performing Arithmetical operations in REPL

Reviewer: Use the markdown header `####`.

Something like this `#### Performing arithmetical operations in REPL`

In the previous example, we printed 'hello world' string on the screen, in this section let's dive in and look at arithmetics:  

In your terminal, start REPL: 

```bash
$ node
> 

```
Let's perform basic addition, subtraction,modulus, division and multipliation respectively:
Output:  

```bash
> 2+2
4
> 5-2
3
> 10%3
1
> 20/2
10
> 10*2
20
> 

```

> 2. Performing operations using Node's Math library

Reviewer: Change header

Math library is an object with defined methods and properties. We can therefore use these features to perform some arithmetics.  

Example:

```bash

> ---skipped lines

Reviewer: Is this a command?

> function getRandomNumber(maximum){
... return Math.floor(Math.random()*maximum);
... }

> undefined

Reviewer: This snippet is unclear. Break it down into 2 different code snippets - one for input and other for output

Author: This is a simple Node.js method, returning a random number, how is it unclear?

Reviewer: @odiwuoramos 
What I meant by unclear is - 

```bash
function getRandomNumber(maximum){
... return Math.floor(Math.random()*maximum);
... }
```
The above snippet is more like an input, and you get the output as shown below:

```bash
undefined
```

Now, you had put them under same code snippet. So, there are high chances that the reader might understand writing `undefined` as an extension to the above function.

That's the reason, I suggest to put them under different snippets like:

```bash
> function getRandomNumber(maximum){
... return Math.floor(Math.random()*maximum);
... }
```

**Output:**

```bash
undefined
```

Or better put a screenshot, it would resolve all the confusions.

Hope it is clear.

```

Output:
```bash
> console.log(getRandomNumber(20));
12
undefined
> 
```
In REPL, we define `getRandomNumber(args)` and pass it a maximum number. The method uses `Math.floor()` to return a random largest integer less than or equal to a given number.

> It's important to note that `Math.random()` does not provide secure random numbers. Do not use them on security-related tasks.

> 3. Calling Methods (Functions) Using REPL

Reviewer: Change header

Normally, we write functions to handle specific tasks. REPL provides an easy way to handle these methods.   
In JavaScript, we commonly use the `console.log()` global method to print messages. 

Let's enter the following in the prompt: 

```bash
$ node
> console.log('Hello world');
```
Output: 

```bash 
Hello world
undefined // this is the return value of this method, console.log()
> 
```

When you press ENTER, the output is displayed as shown above.  

Next, let's create a function to add two numbers:  

```bash
--- skipped lines

> function addTwoNumbers(firstNumber,lastNumber){
... console.log(firstNumber+ lastNumber)
... }
undefined

> ```

Reviewer: Why do we use `...`? And, why do we get `undefined`?

Author: Node.js??????

We've a function `addTwoNumbers(arguments)` which takes two arguments, `firstNumber` and `lastNumber`.
This function, whenever it's called, logs out the addition of these two numbers.  

Output:

```bash
---- skipped lines
> addTwoNumbers(20,40);
60
undefined
> 
```

We call this method and pass it two parameters, `addTwoNumbers(20,40)`( you're free to pass any integer value).
By pressing ENTER, the result is displayed on the screen as shown above.  

> 4. Creating Variables using REPL

Reviewer: Change header

Creating variables works the same as it would in your`.js` file.
Let's look at an example:  

```bash
$node
> let name = 'john doe';
```
Output on calling `name` and pressing ENTER:

```bash
undefined
> name
'john doe'
> 

```
This variable `name` will remain active until you exit the REPL window session. This implies that you can even concatenate this string to another.  
For instance:

```bash
$node
> let name = 'john doe';
undefined
> name
'john doe'
> name+ ' is a student at Stanford';
'john doe is a student at Stanford'
> 

> ```

Reviewer: Almost all the code snippets looks unclear, since there's frequent use of `$` and `>`.

I would suggest to put screenshots of the code snippets instead, which would enhance the readability of the code snippets.

### Conclusion

In this tutorial, we've seen how we can use the interactive Node.js REPL tool environment.

> We've seen how we can perform different tasks without the necessity to create a js file.

Reviewer: Conclusion could have been more elaborate - what are the key takeaways from this tutorial, give a summary of what the reader has learnt.

> Happy Coding!!

Reviewer: If possible, add reference links under `Further reading` section.

### Introduction
Node.js `Read-Evaluate-Print-Loop`(REPL) is an easy-to-use command-line tool for processing Node.js expressions.

> It captures the user's JavaScript code inputs, interprets and evaluates the result of this code, displays the result to the screen, and repeats the process till the user quits the shell.  

Author: @srishilesh, this paragraph here, says ...."it captures (reading), evaluate, display(print) and repeat(loop) what's it missing?.".

### Introduction
Node.js `Read-Evaluate-Print-Loop`(REPL) is an easy-to-use command-line tool for processing Node.js expressions.
It captures the user's JavaScript code inputs, interprets and evaluates the result of this code, displays the result to the screen, and repeats the process till the user quits the shell.  
It's also important to note that this tool does not require file creation to write codes. It comes with ready to use Node.js development environment.

In this tutorial, we learn the basics of Node.js REPL, how we can use this amazing tool to run scripts without creating `.js` files.

> ### Table of contents

Author: I'm quite unable to understand this statement...could you further clarify. It has a header ### matching any other headings..

> ### Table of contents

Reviewer 2: Check out this commit to see what I meant.

99fadc7

Now how do we exit REPL?
To exit this tool, press `Ctrl+C` on your keyboard.  

Output:
```bash
> let hello = 'hello world';
undefined
> hello
'hello world'
> 
(To exit, press Ctrl+C again or Ctrl+D or type .exit)
> 

```
As shown in the output, you can as well use `ctrl+D` or type `.exit`.  

### Executing code using REPL.

As discussed previously, this tool simplifies work by providing a quick way to test Node.js code without creating files.  
> It's also important to note that any valid JavaScript code can be executed using REPL.

#### Performing Arithmetical operations in REPL
In the previous example, we printed 'hello world' string on the screen, in this section let's dive in and look at arithmetics:  

In your terminal, start REPL: 

```bash
$ node
> 

```
Let's perform basic addition, subtraction,modulus, division and multipliation respectively:
Output:  

```bash
> 2+2
4
> 5-2
3
> 10%3
1
> 20/2
10
> 10*2
20
> 

```
#### Performing operations using Node's Math library
Math library is an object with defined methods and properties. We can therefore use these features to perform some arithmetics.  

Example:

```bash

> > function getRandomNumber(maximum){

Reviewer 2: Please split this into two parts as @srishilesh had suggested.

https://github.com/section-engineering-education/engineering-education/pull/2059#discussion_r626542548

... return Math.floor(Math.random()*maximum);
... }
undefined

```

> Output:

Reviewer 2: Please add screenshots of the code snippets for better readability.
The previous reviewer had suggested the same. 
https://github.com/section-engineering-education/engineering-education/pull/2059#discussion_r626490702

Author: We resolved this, no screenshots required as codes are already here.  REPL is different from the normal editor.

Output:

```bash
> addTwoNumbers(20,40);
60
undefined
> 
```

We call this method and pass it two parameters, `addTwoNumbers(20,40)`( you're free to pass any integer value).
By pressing ENTER, the result is displayed on the screen as shown above.  

#### Creating Variables using REPL

Creating variables works the same as it would in your`.js` file.
Let's look at an example:  

```bash
$node
> let name = 'john doe';
```
Output on calling `name` and pressing ENTER:

```bash
undefined
> name
'john doe'
> 

```
This variable `name` will remain active until you exit the REPL window session. This implies that you can even concatenate this string to another.  
For instance:

```bash
$node
> let name = 'john doe';
undefined
> name
'john doe'
> name+ ' is a student at Stanford';
'john doe is a student at Stanford'
> 

```
### Conclusion

In this tutorial, we've seen how we can use the interactive Node.js REPL tool environment. 
We've seen how we can perform different tasks such as basic artihmetic operations and creating functions without the necessity to create a `.js` file.
You can as well follow [this](https://nodejs.dev/learn/how-to-use-the-nodejs-repl) link for more information.

> 

Reviewer 2: Please add a `further reading` section where you include links to resources that the reader can use to build upon his/her knowledge about Node.js and the REPL

### Introduction

> Node.js `Read-Evaluate-Print-Loop`(REPL) is an easy-to-use command-line tool for processing Node.js expressions.

Reviewer 2: The introduction to REPL seems inadequate. Please incorporate the changes as the previous reviewer had mentioned. I have attached a screenshot below.

![image](https://user-images.githubusercontent.com/34317373/118394370-f006ea00-b661-11eb-93fe-ad35be6567da.png)

https://www.geeksforgeeks.org/node-js-repl-read-eval-print-loop/

Refer to the above link to get an overview of how they've explained the REPL.

It captures the user's JavaScript code inputs, interprets and evaluates the result of this code, displays the result to the screen, and repeats the process till the user quits the shell.  
It's also important to note that this tool does not require file creation to write codes. It comes with ready to use Node.js development environment.

In this tutorial, we learn the basics of Node.js REPL, how we can use this amazing tool to run scripts without creating `.js` files.

### Table of contents
1. Introduction to REPL
2. Basic Arithmetic operations using REPL
3. Creating methods and variables using REPL

### Prerequisites

This tutorial requires:

> 1. Node.js Installed in your development local machine.

Reviewer 2: Please include a link to install Node.js

2. Basic knowledge of JavaScript and Node.js.

> 

Reviewer 2: Please include some resources (links to section.io articles) that can help readers learn some fundamentals on Node.js

You can find articles here:
[https://section.io/engineering-education](https://section.io/engineering-education)

### Objectives

> At the end of this tutorial, you should be able to use REPL tool to run your Node.js scripts.

Reviewer 2: In the objectives section, elaborate in 3-4 lines on what we are doing in this article. For example, you can include your h4 subheadings and explain them. This ensures that the reader gets a good overview of what they're gonna be learning.

### Getting started with REPL

Enter the following in the terminal:

```bash
node
```

**Output:**

```bash
> 
```

Entering the command `node` in the terminal starts the REPL command-line tool indicated by the `>` symbol.  
This symbol acts as an indicator that JavaScript is ready to read and evaluate your code.  

We can test the functionality by printing `hello world` using REPL as shown below:

```bash
node
> let hello = 'hello world';
undefined
> hello
'hello world'  // output
> 
```

> In the script, we assign the text `hello world` to a variable called `hello`. Now, on calling this variable `hello`, we get `hello world` as the output on the command shell. You also notice that on pressing ENTER, `undefined` is printed. This is because `hello` variable doesn't return a value.

Reviewer 2: The explanation here is confusing.

> This is because `hello` variable doesn't return a value.

The `hello` variable does return a value. It returns the value: hello world. You have to mention that the declaration of the hello variable,

That is, the line:
```js
let hello = 'hello world';
```
does not return anything and hence returns `undefined`

```bash
> function addTwoNumbers(firstNumber,lastNumber) {
... console.log(firstNumber+ lastNumber)
... }
undefined
```

We've a function `addTwoNumbers(arguments)` which takes two arguments, `firstNumber` and `lastNumber`.
This function, whenever it's called, logs out the addition of these two numbers.  

**Output:**

```bash
> addTwoNumbers(20,40);
60
undefined
> 
```

We call this method and pass it two parameters, `addTwoNumbers(20,40)`( you're free to pass any integer value).
By pressing ENTER, the result is displayed on the screen as shown above.  

#### Creating Variables using REPL

Creating variables works the same as it would in your`.js` file.
Let's look at an example:  

```bash
node
> let name = 'john doe';
```

Output on calling `name` and pressing ENTER:

```bash
undefined
> name
'john doe'
> 
```

This variable `name` will remain active until you exit the REPL window session. This implies that you can even concatenate this string to another. For instance:

```bash
node
> let name = 'john doe';
undefined
> name
'john doe'
> name+ ' is a student at Stanford';
'john doe is a student at Stanford'
> 

```

### Conclusion

> In this tutorial, we've seen how we can use the interactive Node.js REPL tool environment. 

Reviewer 2: Please give a more elaborate conclusion. As the previous reviewer mentioned.
Summarize everything you did in the article (Performing Arithmetical operations in REPL, Performing operations using Node's Math library, Calling Methods (Functions) Using REPL, Creating Variables using REPL)

We've seen how we can perform different tasks such as basic artihmetic operations and creating functions without the necessity to create a `.js` file.  

### Further reading

> Follow [this](https://nodejs.dev/learn/how-to-use-the-nodejs-repl) link for more information.

Reviewer 2: Include more links and articles where the reader can learn more about REPL, REPL alternatives and more.
